### PR TITLE
[Claim] check other chains for claimable amts + click to change networks

### PR DESCRIPTION
**Pull request by W3stside**
# Summary

Shows claims in other chains and allows clicking chain name to change network

Closes #TBD 

*Add screenshots if applicable. Images are nice :)*
What it looks like
image

Can click and change networks:
image

  # To Test

1. <<Step one>> Open the page `about`
- [ ] <<What to expect?>> Verify it contains about information...
- [ ] Checkbox Style list of things a QA person could verify, i.e.
- [ ] Should display Text Input our storybook
- [ ] Input should not accept Numbers
2. <<Step two>> ...

  # Background

*Optional: Give background information for changes you've made, that might be difficult to explain via comments*

### CODE REVIEW

**Review by anxolin — COMMENTED** [review]
I like very much how this idea is implemented, two quick feedbacks:


1. It's telling me I have a claiming available in GC, but this is where i am and I don't see it. 
image

2.  It doesn't make sense to show you have a claiming in the network you are. I would exclude the current network from the message. 

What about:

`You ${hasClaim ?  ' also' : '' } have available claims on ${otherNetworks}`

File: src/custom/hooks/useChangeNetworks.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+  // if wallet supports 3085
+  useEffect(() => {
+    if (queuedNetworkSwitch && account && chainId && implements3085) {
+      networkCallback(queuedNetworkSwitch)
+      setQueuedNetworkSwitch(null)
+    }
+  }, [networkCallback, queuedNetworkSwitch, chainId, account, implements3085])
+
+  return {
+    callback: networkCallback,
```
Comment (anxolin) [inline src/custom/hooks/useChangeNetworks.ts:69]: can you give this a name that would make it clear  what does it do? 
i.e. `switchNetwork`

File: src/custom/hooks/useChangeNetworks.ts [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+import { CHAIN_INFO, SupportedChainId } from 'constants/chains'
+import { switchToNetwork } from 'utils/switchToNetwork'
+import { supportedChainId } from 'utils/supportedChainId'
+
+export default function useChangeNetworks() {
+  const { chainId: preChainId, library, account } = useActiveWeb3React()
+  const { error } = useWeb3React() // MOD: check unsupported network
+  const node = useRef<HTMLDivElement>()
+  const open = useModalOpen(ApplicationModal.NETWORK_SELECTOR)
+  const toggle = useToggleModal(ApplicationModal.NETWORK_SELECTOR)
```
Comment (anxolin) [inline src/custom/hooks/useChangeNetworks.ts:17]: maybe give it a better name? i was in line 36 trying to guess what it toggles

File: src/custom/hooks/useChangeNetworks.ts [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+  useEffect(() => {
+    if (queuedNetworkSwitch && account && chainId && implements3085) {
+      networkCallback(queuedNetworkSwitch)
+      setQueuedNetworkSwitch(null)
+    }
+  }, [networkCallback, queuedNetworkSwitch, chainId, account, implements3085])
+
+  return {
+    callback: networkCallback,
+    conditionalToggle,
```
Comment (anxolin) [inline src/custom/hooks/useChangeNetworks.ts:70]: i saw this, but is not used. Why do we need to expose it?  not saying we shouldn't. Trying to understand

File: src/custom/pages/Claim/ClaimsOnOtherChainsBanner.tsx [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+  }
+`
+
+function ClaimsOnOtherChainsBanner({ className }: { className?: string }) {
+  const { callback } = useChangeNetworks()
+
+  const { hasClaimsOnOtherChains } = useClaimState()
+  const chainsWithClaims: SupportedChainId[] = useMemo(
+    () =>
+      Object.keys(hasClaimsOnOtherChains).reduce((acc, chain) => {
```
Comment (anxolin) [inline src/custom/pages/Claim/ClaimsOnOtherChainsBanner.tsx:49]: can you skip the current network?

Also, the reduce could also return a flag saying if the current one has it?

With these 2 pieces of info, we can do the message i was suggesting.

File: src/custom/state/claim/updater.tsx [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  const { setHasClaimsOnOtherChains } = useClaimDispatchers()
+
+  useEffect(() => {
+    if (!account) return
+
+    const supportedChain = supportedChainId(chainId)
+    SUPPORTED_CLAIM_CHAINS.forEach((chain) => {
+      if (supportedChain === chain) return
+      // Chain at this point is a supported chain
+      fetchClaims(account, chain)
```
Comment (anxolin) [inline src/custom/state/claim/updater.tsx:22]: this doesn't account for the consumed claims.

File: src/custom/state/claim/updater.tsx [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+        .then((userClaims) => {
+          console.debug('[AvailbleClaimsUpdater]::Claim on', chain, userClaims)
+          setHasClaimsOnOtherChains({ chain, hasClaims: true })
+        })
+        .catch((error) => {
+          console.error('[AvailbleClaimsUpdater]::Error for chain', chain, error)
+          setHasClaimsOnOtherChains({ chain, hasClaims: false })
+        })
+    })
+  }, [account, blockNumber, chainId, setHasClaimsOnOtherChains])
```
Comment (anxolin) [inline src/custom/state/claim/updater.tsx:32]: isn't this a performance issue if you add this check on every block? 

This is likely something we would want to check once, right? we then just make when the user claims, we take it into account

**Review by anxolin — COMMENTED** [review]
A bit big too review the code. I realised some of the comments from my review are related to the change network hook, which i later realised is a copy pasta, so not your code. 

Anyways, maybein this case we can change it, since we didn't do a mod, and we just duplicated the logic to own it.

File: src/custom/state/claim/updater.tsx (hunk 6 above)
Comment (W3stside) [reply src/custom/state/claim/updater.tsx:32]: yeah blockNumber here was for testing, will remove

File: src/custom/pages/Claim/ClaimsOnOtherChainsBanner.tsx (hunk 4 above)
Comment (W3stside) [reply src/custom/pages/Claim/ClaimsOnOtherChainsBanner.tsx:49]: shouldnt even be showing the chain we're on, i checked this. likely i messed sth up when testing. updater checks current chain

File: src/custom/hooks/useChangeNetworks.ts (hunk 2 above)
Comment (W3stside) [reply src/custom/hooks/useChangeNetworks.ts:17]: this is all old code i just pasted

File: src/custom/hooks/useChangeNetworks.ts (hunk 1 above)
Comment (W3stside) [reply src/custom/hooks/useChangeNetworks.ts:69]: this is all old code i just pasted

File: src/custom/pages/Claim/ClaimsOnOtherChainsBanner.tsx (hunk 4 above)
Comment (W3stside) [reply src/custom/pages/Claim/ClaimsOnOtherChainsBanner.tsx:51]: fixed

File: src/custom/hooks/useChangeNetworks.ts (hunk 3 above)
Comment (W3stside) [reply src/custom/hooks/useChangeNetworks.ts:70]: i exposed it for use in the original code i need to add the hook to, wihch i do in #2318 
it is changing the logic, tho

File: src/custom/state/claim/updater.tsx (hunk 5 above)
Comment (alfetopito) [reply src/custom/state/claim/updater.tsx:20]: Yeah, you'll have to use `useUserAvailableClaims` hook here although it uses the connected chainId.
Will need to modify the hooks to take it as an optional parameter instead

Comment (W3stside) [reply src/custom/state/claim/updater.tsx:20]: @anxolin https://github.com/gnosis/cowswap/pull/2320 @alfetopito

**Comment by alfetopito** [comment]
Suggestion about UI

The banner is not as "round" as the other styles used on claim UI
Screen Shot 2022-01-26 at 15 54 35

Example
Screen Shot 2022-01-26 at 15 54 50

Also, what about re-using the same banner that tells you the account has claims for this?

Instead of the checkmark, use an exclamation and the same text.

**Comment by elena-zh** [comment]
In addition to @alfetopito comments above, I'd like to propose change text banner to 'This account has available claims in <>'  due to 'You have..' sounds a bit illogical when chect available claims on behalf of someone else. 
image

**Comment by elena-zh** [comment]
Some other issues: 
1. It would be nice to remove link from 'and'
image
2. Make banner's text center-aligned?
image

3. Remove links from the banner when it is impossible to switch to another network (wallet does not support OR connected using WC): 
image

4. Maybe we could show this banner outside the Claim area like we do for the affiliate flow? 
image
image

File: src/custom/state/claim/updater.tsx (hunk 6 above)
Comment (anxolin) [reply src/custom/state/claim/updater.tsx]: Okay, lets make sure we don't add more performance issues 👌

File: src/custom/pages/Claim/ClaimsOnOtherChainsBanner.tsx [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    <PhishAlert isActive className={className}>
+      <Wrapper>
+        <div>You have available claims on</div>
+        <div>
+          {chainsWithClaims.map((chainId, index, array) => {
+            const changeNetworksCallback = () => callback(chainId)
+            const isLastInMultiple = index === array.length - 1 && array.length > 1
+            return (
+              <ChainSpan key={chainId} onClick={changeNetworksCallback}>
+                {`${isLastInMultiple ? 'and ' : ''}${NETWORK_LABELS[chainId]}`}
```
Comment (anxolin) [inline src/custom/pages/Claim/ClaimsOnOtherChainsBanner.tsx:76]: I would follow the tip from Elena on not adding this into the link

**Review by anxolin — APPROVED** [review]
This is a great feature! works very well now. 

I agree with all design suggestions. This PR is also big enough, should we merge and do a follow up?

**Comment by biocom** [comment]
Agree on merging to follow up. My feedback on @elena-zh 's points:
1) Agreed.
2) Agreed (can address).
3) Makes sense, agreed.
4) I would show it inside the container. In my opinion that's contextually the most relevant position.

**Comment by elena-zh** [comment]
> In addition to @alfetopito comments above, I'd like to propose change text banner to 'This account has available claims in <>' due to 'You have..' sounds a bit illogical when check available claims on behalf of someone else. image

Will we address this?

**Comment by elena-zh** [comment]
I have just retested changes when connected to Gnosis Safe: is it possible to have safe with the same address in all networks? 
image

If not, I'd rather not to show this message when connected to Gnosis safe

**Comment by W3stside** [comment]
> I have just retested changes when connected to Gnosis Safe: is it possible to have safe with the same address in all networks? image
> 
> If not, I'd rather not to show this message when connected to Gnosis safe

lets reference in another issue please

**Comment by elena-zh** [comment]
@W3stside , I reported it in #2333

**Pull request merged**